source/common/http/conn_manager_impl.cc

```diff
@@ -1099,8 +1155,16 @@ void ConnectionManagerImpl::ActiveStream::encodeData(ActiveStreamEncoderFilter*
                    end_stream);
 
   request_info_.addBytesSent(data.length());
-  response_encoder_->encodeData(data, end_stream);
-  maybeEndEncode(end_stream);
+
+  // If trailers were adding during encodeData we need to trigger decodeTrailers in order
+  // to allow filters to process the trailers.
+  if (end_stream && response_trailers_) {
+    response_encoder_->encodeData(data, false);
+    encodeTrailers(filter, *response_trailers_);
```

Comment:
Same comment here about keeping track of which filter added the trailers?

As an aside, I'm a little confused as to why the logic in encodeHeaders() does not mirror decodeHeaders() WRT to handling continue/last filter. (And yes I know I wrote this code!) I can look into this more tomorrow or Friday, but if you feel like taking a look that would be appreciated.

Reply:
I looked into it a little bit: adding the same last filter check to encodHeaders causes a test failure in `HttpConnectionManagerImplTest.HitRequestBufferLimitsIntermediateFilter` because of a missing expectation for a call to `response_encoder.encodeHeaders`, and an integration test fails due to 
```
[2018-08-09 23:36:14.552][694871][critical][assert] bazel-out/darwin-dbg/bin/source/common/request_info/_virtual_includes/request_info_lib/common/request_info/request_info_impl.h:83] assert failure: !first_downstream_tx_byte_sent_.
```

which presumably comes from hitting this code more than once (towards the end of `encodeHeaders`):

```
  // Now actually encode via the codec.
  request_info_.onFirstDownstreamTxByteSent(); // <- this bit
  response_encoder_->encodeHeaders(headers,
                                   end_stream && continue_data_entry == encoder_filters_.end());

```

The test that's failing is `IpVersions/Http2IntegrationTest.HittingDecoderFilterLimit/IPv4` - I'm guessing the filter used there normally stops iteration, but by continuing on the last filter we end up running through the last part of `encodeHeaders` again. Seems like the issue is that somehow  `encodeTrailers` is being called twice, and special casing the last filter causes the second call to hit `onFirstDownstreamTxByteSent`.

Unclear whether this is *why* we don't special case the last filter in `encodeHeaders`, or if it's just a issue that hasn't surfaced because we don't.

Reply:
OK, thanks for looking. I can take a look tomorrow or this weekend more. I wouldn't worry about it for now but it seemed strange to me that it wasn't symmetrical and I couldn't remember why.